ingress/ingress.go

```diff
@@ -121,12 +125,29 @@ func (c *Controller) syncIngress(key string) error {
 			return nil
 		}
 	}
+	rollouts, err := c.getRolloutsByIngress(ingress.Namespace, ingress.Name)
+	if err != nil {
+		return nil
+	}
+	// An ingress without annotations cannot be a alb or nginx ingress
+	if ingress.Annotations == nil {
+		return nil
+	}
 
-	if rollouts, err := c.getRolloutsByIngress(namespace, name); err == nil {
-		for i := range rollouts {
-			// reconciling the Rollout will ensure the canaryIngress is updated or created
-			c.enqueueRollout(rollouts[i])
-		}
+	switch ingress.Annotations["kubernetes.io/ingress.class"] {
+	case "aws-alb":
+		return c.syncALBIngress(ingress, rollouts)
+	case "nginx":
```

Comment:
Does this mean that all Ingress configs must have the kubernetes.io/ingress.class annotation to work with traffic shaped rollouts? How would this work for clusters with multiple nginx ingresses with different names or don't use the name?

In our large shared cluster we actually have multiple nginx ingress for different envs and service requirements, with this restriction all those would be excluded from using this feature.

Reply:
Hey, @cronik that's a great point! I'm going to add some command-line flags that allow you to specify which values of kubernetes.io/ingress.class are used for nginx and albs

Reply:
I also just saw that users can change their annotation prefix (i.e. `alb.ingress.kubernetes.io`) if they want so I'll add that functionality too

Reply:
Okay, I added the `--alb-ingress-classes` and `--nginx-ingress-classes` flags to address these usecases.

Reply:
would it also make sense to add an option like `--default-ingress-class` for cases where it is not annotated explicitly?

Reply:
I.E if there's no "kubernetes.io/ingress.class" annotation, have the flag dicate that the controller should consider it an NGINX or ALB ingress? Do the ingress controller support that behavior?

Reply:
Yeah, I don't know about ALB, but nginx-ingress has this behavior by default unless the ingress class name is explicitly set on the ingress controller.

https://kubernetes.github.io/ingress-nginx/user-guide/multiple-ingress/#multiple-ingress-nginx-controllers
https://kubernetes.github.io/ingress-nginx/user-guide/cli-arguments/

| arg | Description |
| --- | ------------ |
| `--ingress-class string` | Name of the ingress class this controller satisfies. The class of an Ingress object is set using the annotation "kubernetes.io/ingress.class". All ingress classes are satisfied if this parameter is left empty. |

Basically when `--ingress-class` is not set it acts as the default controller for all ingress.
Our cluster has multiple nginx ingress controllers, most ingress don't specify the ingress class annotation and get picked up by the default controller, when an ingress needs a specific class then the annotation is added to restrict to a specific one.

Reply:
I tested Jesse's solution made in another comment successful (`--nginx-ingress-classes nginx --nginx-ingress-classes '')`. Since there is no good way to pick a default value, I think it is better to leave it unset, and allow users to specify it themselves if needed.